Mathlib/Data/Matrix/Basis.lean

```diff
@@ -216,6 +216,44 @@ theorem mul_of_ne {k l : n} (h : j ≠ k) (d : α) :
 
 end
 
+section Commute
+
+variable [Fintype n]
+
+theorem comm_stdBasisMatrix {i j k : n} {M : Matrix n n α}
+    (hM : Commute (stdBasisMatrix i j 1) M) (hkj : k ≠ j) : M j k = 0 := by
+  have := Matrix.ext_iff.mpr hM i k
+  aesop
+
+theorem comm_stdBasisMatrix' {i j k : n} {M : Matrix n n α}
+    (hM : Commute (stdBasisMatrix i j 1) M) (hki : k ≠ i) : M k i = 0 := by
+  have := Matrix.ext_iff.mpr hM k j
+  aesop
+
+theorem comm_stdBasisMatrix'' {i j : n} {M : Matrix n n α}
+    (hM : Commute (stdBasisMatrix i j 1) M) : M i i = M j j := by
+  have := Matrix.ext_iff.mpr hM i j
+  aesop
+
+/-- `M` is a scalar matrix if it commutes with every non-diagonal `stdBasisMatrix`.​-/
+theorem comm_all_stdBasisMatrix_nondiag [Inhabited n] {M : Matrix n n α}
+    (hM : ∀ (i j : n), i ≠ j → Commute (stdBasisMatrix i j 1) M) :
+    M = M default default • (1 : Matrix n n α) := by
```

Comment:
I think a better statement would be that `M \in Set.range Matrix.scalar` or similar, as that has the advantage of not needing `Inhabited n`

Reply:
Unfortunately, that just turns `Inhabited n` into `Inhabited α`. I think it's fine to assume `Inhabited n` directly.  As long as we have `Nonempty n`, we can simply use `Classical.inhabited_of_nonempty ‹_›` to obtain `Inhabited n`. Moreover, usually we use `Fin n` where n > 0, in which case `Inhabited (Fin n)` holds automatically since there is an instance in `Mathlib/Logic/Unique.lean`.

Reply:
Now I have restated it and removed `Inhabited n`.

Reply:
[](url)

> Unfortunately, that just turns `Inhabited n` into `Inhabited α`.

I don't know what you mean by this. You shouldn't need `Inhabited α`, you can just use `1` or `0`.

Reply:
@eric-wieser I thought you wanted me to derive `Nonempty n` from `M : Matrix n n α`.
In this PR, it was initially expressed as `∃ (c : α), M = c • (1 : Matrix n n α)`, which is not much different from `M ∈ Set.range (Matrix.scalar n)`. Later, I changed it to `M = M default default • (1 : Matrix n n α)` to make it easier to use in the tactic `rewrite`.
To remove the assumption of `Inhabited n`, I changed it to `(i : n) : M = M i i • (1 : Matrix n n α)`. This is equivalent to `∀ (i : n), M = M i i • (1 : Matrix n n α)`, and it holds even if `n` is empty. When `n` is non-empty, I believe this form provides quite precise information and is more user-friendly.
In my opinion, when we want to handle the case where `n` is empty, we should directly jump to `M = 1`, rather than trying invoke this theorem. From this perspective, I think even `Inhabited n` is acceptable.
I believe that `(i : n) : M = M i i • (1 : Matrix n n α)` is a better expression. But I agree that the theorem could be renamed to `mem_range_scalar_of_commute_stdBasisMatrix`.

Reply:
Oh! The case where `n` is empty and `α` is non-empty is really confusing! There should be a theorem that tells us `Matrix n n α` is isomorphic to `α` in this case.

Reply:
> There should be a theorem that tells us `Matrix n n α` is isomorphic to `α` in this case.

I find that hard to believe, given such a theorem is false!

Reply:
> Later, I changed it to `M = M default default • (1 : Matrix n n α)` to make it easier to use in the tactic `rewrite`.

My claim is that this is actually less useful, as it contains `M` on both sides so is hard to rewrite by. The `Set.range` version can be used with `cases`, and it will substitute all the `M`s globally for some scalar `r`

Reply:
> The `Set.range` version can be used with `cases`, and it will substitute all the `M`s globally for some scalar `r`

Fine. I will get used to it.

Reply:
> It's isomorphic to `Unit` instead.

Now it gets even more counter-intuitive! Mathlib should add this theorem.

Reply:
> I find that hard to believe, given such a theorem is false! It's isomorphic to `Unit` instead.

Wait, what do you think of `subsingleton_of_empty_left` in `Mathlib/Data/Matrix/Basic.lean`?

Reply:
That tells us that when the index type is empty, then all matrices are `0`